src/StrawberryShake/Client/src/Transport.WebSockets/SocketClient.cs

```diff
@@ -0,0 +1,239 @@
+
+        private ISocketProtocol? _activeProtocol;
+        private bool _disposed;
+        private readonly string _name;
+
+        public WebSocketClient(string name, IReadOnlyList<ISocketProtocol> protocols)
+        {
+            _protocols = protocols;
+            _name = name;
+            _socket = new ClientWebSocket();
+
+            for (var i = 0; i < _protocols.Count; i++)
+            {
+                _socket.Options.AddSubProtocol(_protocols[i].ProtocolName);
+            }
+        }
+
+        public Uri? Uri { get; set; }
+
+        public string? Name => _name;
+
+        public bool IsClosed =>
+            _disposed
+            || _socket.CloseStatus.HasValue;
+
+        public async Task OpenAsync(CancellationToken cancellationToken = default)
+        {
+            if (_disposed)
+            {
+                throw new ObjectDisposedException(nameof(WebSocketClient));
+            }
+
+            if (Uri is null)
+            {
+                // TODO: Uri should not be null
+                throw new InvalidOperationException();
+            }
+
+            await _socket.ConnectAsync(Uri, cancellationToken).ConfigureAwait(false);
+
+            for (var i = 0; i < _protocols.Count; i++)
+            {
+                if (_protocols[i].ProtocolName == _socket.SubProtocol)
+                {
+                    _activeProtocol = _protocols[i];
+                    break;
+                }
+            }
+
+            if (_activeProtocol is null)
+            {
+                await CloseAsync(
+                        "Failed to initialize protocol",
+                        SocketCloseStatus.ProtocolError,
+                        cancellationToken)
+                    .ConfigureAwait(false);
+
+                // TODO throw error
+                throw new InvalidOperationException();
+            }
+
+            await _activeProtocol.InitializeAsync(this, cancellationToken).ConfigureAwait(false);
+        }
+
+        public async Task CloseAsync(
+            string message,
+            SocketCloseStatus closeStatus,
+            CancellationToken cancellationToken = default)
+        {
+            try
+            {
+                if (IsClosed)
+                {
+                    return;
+                }
+
+                if (_activeProtocol is not null)
+                {
+                    await _activeProtocol.TerminateAsync(cancellationToken).ConfigureAwait(false);
+                }
+
+                await _socket.CloseOutputAsync(
+                        MapCloseStatus(closeStatus),
+                        message,
+                        cancellationToken)
+                    .ConfigureAwait(false);
+
+                await DisposeAsync();
+            }
+            catch
+            {
+                // we do not throw here ...
+            }
+        }
+
+        public ISocketProtocol GetProtocol()
+        {
+            if (_activeProtocol is null)
```

Comment:
Smells 

this check if needed to see if the protocol is initialized corrreclty

Reply:
probably some instance variable would be better

Reply:
best would be if this is not possible in first place